commit fe6fb38b3d4095351955b9872d0fd6cba64f8784
Author: Mark Vitale <mvitale@sinenomine.net>
Date:   Fri Nov 30 15:13:24 2018 -0800

    Linux 4.20: disable -settime support
    
    In Linux v4.20, do_settimeofday() has been removed in favor of
    do_settimeofday64().
    
    This commit is for 1.6.x only.  On master and 1.8.x, -settime support
    has been removed from OpenAFS, but on 1.6.x some sites may still be
    specifying -settime even though it is deprecated and known to be broken.
    Therefore, for 1.6.x we can't simply remove afs_osi_SetTime.  It's also
    a waste of time to convert it to use do_settimeofday64() because the
    -settime feature is deprecated.
    
    Instead, add an autoconf test to detect the absence of
    do_settimeofday().  When do_settimeofday() is missing and -settime has
    been specified, log a warning message whenever OpenAFS attempts to
    adjust the system time:
    
       'afs attempted to set clock; use "afsd -nosettime"'
    
    do_settimeofday() was introduced sometime during v2.4.x and thus has
    always been present in all Linux versions covered by OpenAFS conditional
    AFS_LINUX26_ENV (until now).
    
    We don't need to do anything with the LINUX24 version of
    afs_osi_SetTime() since do_settimeofday() will never disappear on those
    old Linux versions.
    
    Change-Id: Ia147406e7e2cf42f76466c85b0392a8559bd112b
    Reviewed-on: https://gerrit.openafs.org/13403
    Tested-by: BuildBot <buildbot@rampaginggeek.com>
    Reviewed-by: Cheyenne Wills <cwills@sinenomine.net>
    Reviewed-by: Marcio Brito Barbosa <mbarbosa@sinenomine.net>
    Reviewed-by: Stephan Wiesand <stephan.wiesand@desy.de>

diff --git a/src/afs/LINUX/osi_misc.c b/src/afs/LINUX/osi_misc.c
index 50dd625e9..5490c5dca 100644
--- a/src/afs/LINUX/osi_misc.c
+++ b/src/afs/LINUX/osi_misc.c
@@ -31,13 +31,20 @@ afs_ucred_t afs_osi_cred;
 void
 afs_osi_SetTime(osi_timeval_t * tvp)
 {
+#ifdef HAVE_LINUX_DO_SETTIMEOFDAY
     struct timespec tv;
     tv.tv_sec = tvp->tv_sec;
     tv.tv_nsec = tvp->tv_usec * NSEC_PER_USEC;
 
     AFS_STATCNT(osi_SetTime);
-
     do_settimeofday(&tv);
+#else
+    /*
+     * Don't attempt to support -settime on Linux versions that no longer
+     * have 32-bit time APIs.
+     */
+    afs_warn("afs attempted to set clock; use \"afsd -nosettime\"\n");
+#endif
 }
 
 void
diff --git a/src/cf/linux-kernel-func.m4 b/src/cf/linux-kernel-func.m4
index 62fd528d6..72870bfe6 100644
--- a/src/cf/linux-kernel-func.m4
+++ b/src/cf/linux-kernel-func.m4
@@ -39,6 +39,9 @@ AC_CHECK_LINUX_FUNC([d_count],
 AC_CHECK_LINUX_FUNC([d_make_root],
                     [#include <linux/fs.h>],
                     [d_make_root(NULL);])
+AC_CHECK_LINUX_FUNC([do_settimeofday],
+                    [#include <linux/time.h>],
+                    [do_settimeofday(NULL);])
 AC_CHECK_LINUX_FUNC([do_sync_read],
                     [#include <linux/fs.h>],
                     [do_sync_read(NULL, NULL, 0, NULL);])
